consensus/src/dag/bootstrap.rs

```diff
@@ -210,15 +217,27 @@ impl DagBootstrapper {
                     let _ = df_handle.await;
                     return;
                 },
-                certified_node_msg = handler.run(&mut dag_rpc_rx) => {
+                sync_status = handler.run(&mut dag_rpc_rx) => {
                     df_handle.abort();
                     let _ = df_handle.await;
 
-                    let dag_fetcher = DagFetcher::new(self.epoch_state.clone(), self.dag_network_sender.clone(), self.time_service.clone());
+                    match sync_status {
+                        StateSyncStatus::NeedsSync(certified_node_msg) => {
+                            let dag_fetcher = DagFetcher::new(self.epoch_state.clone(), self.dag_network_sender.clone(), self.time_service.clone());
 
-                    if let Err(e) = sync_manager.sync_dag_to(&certified_node_msg, dag_fetcher, dag_store.clone()).await {
-                        error!(error = ?e, "unable to sync");
+                            if let Err(e) = sync_manager.sync_dag_to(&certified_node_msg, dag_fetcher, dag_store.clone()).await {
+                                error!(error = ?e, "unable to sync");
+                            }
+                        },
+                        StateSyncStatus::EpochEnds => {
+                            // Wait for epoch manager to signal shutdown
+                            _ = shutdown_rx.await;
+                            return;
+                        },
```

Comment:
Not sure this is correct, because this peer could have received a old ledger info and it may already be in the new epoch.

Reply:
the invariant we should hold is that the dag instance only receives message from the same epoch. if it's already in the new epoch, it shouldn't receive message from previous epoch.

Reply:
sending multiple/stale epoch change proof is okay I think, epoch manager should handle that

Reply:
We don't check the ledger info is from the same epoch actually. I need to fix that.

Reply:
I added the check to make sure its same epoch.